### ISSUE TITLE
Input Validation example

### ISSUE THREAD

**Issue by PhilippS93** [graphql-java-kickstart/graphql-java-tools#385]
Hello,

Thank you for your work on this library. This is very helpful.

I want to use a directive to validate input fields like so:

input UserInput {
  name: String @Size( min : 2, max : 255)
}

There is a callback in SchemaDirectiveWiring:

```
default GraphQLInputObjectField onInputObjectField(SchemaDirectiveWiringEnvironment<GraphQLInputObjectField> environment) {
        return environment.getElement();
    }
```

but there is no example on how to use this method to validate the input. There are only [examples](https://github.com/graphql-java-kickstart/samples/tree/master/directives/src/main/java/directives) for [onArgument](https://github.com/graphql-java-kickstart/samples/blob/master/directives/src/main/java/directives/RangeDirective.java) and [onField](https://github.com/graphql-java-kickstart/samples/blob/master/directives/src/main/java/directives/UppercaseDirective.java) callbacks.

Is it possible to provide examples for this kind of validation?

Thanks

**Comment by sumitbadaya27** [comment]
thanks a lot. hoping to have it merged soon

**Comment by sumitbadaya27** [comment]
Whoa! great.
so maven version updated? 
Or how do I get the updated version?

**Comment by oliemansm** [comment]
Use snapshot version 7.0.2-SNAPSHOT of graphql-spring-boot-starter. You have to add a snapshot repository to be able to use it, see: https://github.com/graphql-java-kickstart/graphql-spring-boot#snapshots

**Comment by sumitbadaya27** [comment]
After using snapshot version of spring boot starter,  suddenly my GraphQLResolver.java and such classes are not found. "Cannot resolve symbol 'GraphQLResolver" getting this error.
Can you please guide as to how to use this update that you made for validations and input objects?


My POM.XML dependency is as below:
```

<repositories>
        <repository>
            <id>jfrog-snapshots</id>
            <name>oss-jfrog-artifactory-snapshots</name>
            <url>https://oss.jfrog.org/artifactory/oss-snapshot-local</url>
        </repository>
    </repositories>

        <dependency>
            <groupId>com.graphql-java-kickstart</groupId>
            <artifactId>graphiql-spring-boot-starter</artifactId>
            <version>7.0.1</version>
            <scope>runtime</scope>
        </dependency> 
```

**Comment by oliemansm** [comment]
In recent versions base packages have been migrated as stated in the release notes. Remove the imports and your IDE should already help you importing the correct versions (graphql.kickstart based packages).

No you don't include graphql-java-tools directly. Rely on graphql-spring-boot-starter to import graphql-java-tools. In the POM you've described above you're importing GraphiQL, not graphql-spring-boot-starter.

**Comment by sumitbadaya27** [comment]
My pom file content is as below:
```

<repositories>
        <repository>
            <id>jfrog-snapshots</id>
            <name>oss-jfrog-artifactory-snapshots</name>
            <url>https://oss.jfrog.org/artifactory/oss-snapshot-local</url>
        </repository>
    </repositories>

<dependency>
            <groupId>com.graphql-java</groupId>
            <artifactId>graphql-java</artifactId>
            <version>14.0</version>
        </dependency>

<dependency>
            <groupId>com.graphql-java-kickstart</groupId>
            <artifactId>graphql-spring-boot-starter</artifactId>
            <version>7.0.1</version>
        </dependency>
        <dependency>
            <groupId>com.graphql-java-kickstart</groupId>
            <artifactId>graphiql-spring-boot-starter</artifactId>
            <version>7.0.1</version>
            <scope>runtime</scope>
        </dependency>

```


but still somehow kickstart is using 6.0.2 version of GraphQLResolver 


I have performed mvn clean install. Can you help please?

**Comment by oliemansm** [comment]
Come on, just read.... This is costing me a lot of time that's totally unnecessary. I said to use version 7.0.2-SNAPSHOT of graphql-spring-boot-starter and you're clearly using 7.0.1. Don't pull in graphql-java either, that's done for you by graphql-spring-boot-starter too.

```xml
<dependency>
    <groupId>com.graphql-java-kickstart</groupId>
    <artifactId>graphql-spring-boot-starter</artifactId>
    <version>7.0.2-SNAPSHOT</version>
</dependency>
```

**Comment by sumitbadaya27** [comment]
ty

**Comment by sumitbadaya27** [comment]
Dont mean to bother you but now its:

ClassNotFoundException: graphql.schema.GraphQLNamedType


I don't mean to keep this on priority and am sorry for bothering too much. 
Posting this error just so that you have in mind that there could still be some issue in the recent update. Will wait for next stable release. Thanks a lot for the updates though.

**Comment by oliemansm** [comment]
Then try to add the graphql-java dependency anyway. You should start trying to investigate, debug and figure out why certain errors pop-up instead of asking questions each time you run into something. Programming isn't copy pasting code around until something just magically works. It's about understanding how and why something works and applying that. In the end and also in this case you'll only get that kind of understanding by following the code.

If everything else in this library is working as it should and the problem with that type not being found is limited to your use case and configuration then there's no reason to assume it'll be solved by a next release. Might just be caused by that library you're trying to use.

I'm not using directives on input objects, and I'm not using that extended validator directive library either. And I'm not getting paid for the time I put in this project, so there's no incentive whatsoever for me to put in these hours in my free weekend. I'm going to limit my time now to just fixing this original input validation example, and not to the support of that library. If you keep running into problems with that library icw this library then I suggest you figure out the cause and if it's related to a bug or something in this project create a PR to fix it.

**Comment by sumitbadaya27** [comment]
I have figured it out already and fixed it. Was about to post message but was afraid you'd not like it to see another message    .
thanks for the fixes and support till now.

**Comment by oliemansm** [comment]
Can you leave the cause and solution here in a comment then so anybody else that runs into the same problem you ran into will know how to solve it?

**Comment by brendamckenne** [comment]
Sure, 
For anyone who would be using Java-extended-validation library along with kickstart graphql Library,
below is the new updated pom dependencies (versions specifically)
```

 <dependency>
            <groupId>com.graphql-java</groupId>
            <artifactId>graphql-java</artifactId>
            <version>14.0</version>
        </dependency>
       <!-- <dependency>
            <groupId>com.graphql-java-kickstart</groupId>
            <artifactId>graphql-java-tools</artifactId>
            <version>6.0.2</version>
        </dependency>-->
        <dependency>
            <groupId>com.graphql-java-kickstart</groupId>
            <artifactId>graphql-spring-boot-starter</artifactId>
            <version>7.0.2-SNAPSHOT</version>
        </dependency>
        <dependency>
            <groupId>com.graphql-java-kickstart</groupId>
            <artifactId>graphiql-spring-boot-starter</artifactId>
            <version>7.0.2-SNAPSHOT</version>
            <scope>runtime</scope>
        </dependency>

        <dependency>
            <groupId>com.graphql-java</groupId>
            <artifactId>graphql-java-extended-validation</artifactId>
            <version>0.0.3</version>
        </dependency>
```

Note that <artifactId>graphql-java-tools</artifactId> dependency has been commented out. So don't forget to remove/comment out this dependency in case you have it in your pom file.

The reason is that this dependency will somehow fetch old version of jars of GraphQL base packages, which doesn't work with directives validations properly.

With **7.0.2-SNAPSHOT** version, the validations/directives on Input field objects should work fine then. 

Also don't delete/remove the dependency 
```
 <dependency>
            <groupId>com.graphql-java</groupId>
            <artifactId>graphql-java</artifactId>
            <version>14.0</version>
        </dependency>
```

Otherwise your GraphQL Base Packages will fail to load, even though 'graphql-spring-boot-starter' automatically implicitly provides graphql-java dependencies.


So overall its all now in the version changes and dependencies which has brought the fixes and solutions to validaitons/directives.

Hope this helps others! 
Special thanks to @oliemansm  for taking these things on priority and providing quick turnaround on this.

**Comment by oliemansm** [comment]
Thanks @brendamckenne. You might as well delete that graphql-java-tools from your POM entirely btw and just let graphql-spring-boot-starter handle that dependency. That prevents compatibility issues that might otherwise easily occur. We try to update graphql-spring-boot-starter in sync with graphql-java-tools.

**Comment by oliemansm** [comment]
@PhilippS93 Back to your original question. I've updated the sample (not the documentation). Take a look at this [RangeDirective](https://github.com/graphql-java-kickstart/samples/blob/master/directives/src/main/java/directives/RangeDirective.java). It works on both a regular argument and input object. The solution is based on [graphql-java-extended-validation](https://github.com/graphql-java/graphql-java-extended-validation) from the guys that develop graphql-java, the library that provides the directive implementation in the first place.

Documentation on their end is lacking too for this particular use case. Basically how it works is you don't override the `onArgument` or `onInputObjectField` methods, but you override the `onField` method. Then you iterate over the arguments of that field checking if your directive applies or not. You need to walk up over input object types as well in that case and take lists into account if you need to (this sample doesn't do that btw).

Basically you need to write quite a bit of code to get the result of what you'd want. That's because graphql-java provides an API trying to give you us much flexibility as possible, but in this case making it a bit more difficult to work with. 

Perhaps it would be possible to come up with an interface in graphql-spring-boot-starter to simplify it.

**Comment by brendamckenne** [comment]
@oliemansm  just now realized that the validation doesn't work on "Input" type and works only on "type" in schema. Means it is not working on nested input objects.

However this could be an issue with graphql-extended-validation library as am not sure which library is having the issue.

**Comment by PhilippS93** [comment]
Thank you @oliemansm . I have tested the new code but unfortunately it does not work for my case (see below for an example):

```
directive @Size(min : Int = 0, max : Int = 2147483647) on ARGUMENT_DEFINITION | INPUT_FIELD_DEFINITION

input NameInput {
    forename: String! @Size(min : 3, max : 25)
    surname: String! @Size(min : 3, max : 25)
}

input ChangeUserInput {
    name:NameInput
}

type Mutation {
    changeUser(input: ChangeUserInput): User!
}
```

It seems that only ChangeUserInput is considered in the onField method, but not NameInput.

**Comment by oliemansm** [comment]
My example code doesn't traverse all the way up into nested input objects, so you'd have to add that kind of logic to be able to get that working. As I've indicated above it isn't the most developer friendly solution and some kind of convenience interface for this would come in handy to make it easier to work with.

I'll reopen the issue and rename it for nested input directives.

**Comment by dnagarajan89** [comment]
@oliemansm I tried using the range directive logic for an InputObjectType like this. 

type Mutation {
    addBook(bookInput: BookInput**!**) : Book!
}

input BookInput {
    id: Int! @range(minimum: 4, maximum:10)
    name: String!
}

As I have used non null('!') for InputObjectType in the mutation, while unwrapping the non-null I am getting a GraphQLTypeReference instead of GraphQLInputObjectType in this LOC https://github.com/graphql-java-kickstart/samples/blob/master/directives/src/main/java/directives/RangeDirective.java#L38. Is there a way to get the GraphQLInputObjectType from GraphQLTypeReference to navigate the arguments and look for directive ? 

If I use the mutation parameter without non null, the logic is working fine. 

**addBook(bookInput: BookInput)** 

Please let me know if there is a way around.

**Comment by brittany-alves** [comment]
Hello! I'm trying to follow the `RangeDirective` example that was added for input object fields as part of this issue. I'm trying to do something similar myself, but the `SchemaDirectiveWiring.onField` callback is not being hit for input object fields in my own testing (only "works" when I override `SchemaDirectiveWiring.onInputObjectField`, but then we run into the same issue that @PhilippS93 encountered).  

I ran the sample [DirectiveApplication](https://github.com/graphql-java-kickstart/samples/blob/master/directives/src/main/java/directives/DirectiveApplication.java) and confirmed that `onField` wasn't being hit for the `InputObject` fields. Also found [this issue](https://github.com/graphql-java-kickstart/graphql-java-tools/issues/731) reporting that the RangeDirective example isn't working for input object fields. 

Any guidance here would be much appreciated, maybe I'm missing something here! Thanks folks. :)
